[RF] ProcessTimer for logging timings in `RooFit::MultiProcess` and `RooFit::TestStatistics`, and HeatmapAnalyzer for analysing them

Zeff020 opened the pull request:
# This Pull request:

## Changes or fixes:

This MR contains mostly two classes: 
- `RooFit::MultiProcess::ProcessTimer` which keeps a set of potentially overlapping timers per process that can be started and stopped by referring to their identifier string. Timings are written out in json format. Some extra functionality is added to write out at a specified interval.
- `RooFit::MultiProcess::HeatmapAnalyzer` which translates the timings output in log format by the process timer to a heatmap which it can return in the form of a TH2I with the correctly labeled x and y axes. Some extra functionality is available for inspecting the log files and retrieving its metadata.

Tests are included for both classes.

Note that also the interface to enable timing logging is included in the `RooMinimizer::Config` object which can be given to the RooMinimizer and on RooAbsPdf::fitTo. The variable that centrally controls whether timings are logged or not is `RooMinimizer::Config::logTimings_`. 

Also, an initial timing setup is included, with some `ProcessTimer::start_timer` and `ProcessTimer::end_timer` calls throughout the `RooFit::TestStatistics` code.

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

This PR fixes #

guitargeek commented:
@phsft-bot build

Zeff020 commented:
Hi @guitargeek, I've fixed the compiler warning that it was showing before and also resolved a merge conflict that was still present. Could you try building again?

The merge conflict had nothing to do with changes I made in this branch though so I just set it to the way it was present on master. It was related to some change you made a month ago. Could you please check to ensure that that is how you intended it? It is just a single line.

guitargeek commented:
Hi @Zeff020, thanks for the PR! Yes, you resolved the merge conflicts correctly.

But the problem is still tat you resolved them by merging `master` in your MR branch, meaning that your MR branch is now not a trivial fast-forward branch anymore. The ROOT repository settings only allow fast-forward branches to be merged, and also the CI can only deal with fast forward branches. IMO this is done for good reasons: if you first merge `master` into your branch and then your branch is merged back into master, with a potential merge commit that resolves conflicts in every step, the git history becomes some complicated zipped mess. This is also something you should keep in mind for development with git in general.

That means you should not resolve merge conflicts by merging `master`, but by rebasing your PR branch on top of master with `git rebase`. Let's say you want to rebase the last 8 commits of your development branch `processtimer`, the command would be:
```
git rebase --onto master HEAD~8 processtimer
```
Where `master` is now a local branch that corresponds to the current `root-project/master`.

And then you are by git to fix the conflicts in the rebasing process.

Unfortunately, you can't do this anymore now because you have merged master into your branch, so your developments are not the last n commits of the branch anymore. I would maybe create a new branch form current `master`, cherry-pick your original commits on top of it with `git cherry-pick`, and then change the name of the new branch to `processtimer` again such that when you force push, the PR will be updated.

Sorry for this overly complicated stuff!

Zeff020 commented:
Hi @guitargeek, no worries. And thanks for the comment! This time actually GitHub had a "resolve conflicts" button in the browser that I used but this made the merge commit. I agree, it is not so nice for the commit history. Anyway, I will do as you suggested and fix the commit history locally.

guitargeek commented:
Hi @zeff020, little reminder that the conflicts here still need to be resolved

Zeff020 commented:
@guitargeek conflicts resolved

guitargeek commented:
@phsft-bot build

Zeff020 commented:
Hi @guitargeek, do I need to look into these windows errors? Do we expect this to work on windows?

bellenot commented:
@Zeff020  Multiprocess is not supported on Windows (see [RootBuildOptions.cmake#L352-L355](https://github.com/root-project/root/blob/master/cmake/modules/RootBuildOptions.cmake#L352-L355)). The relevant code must be protected and not cause any compilation error

guitargeek commented on roofit/roofitcore/src/RooGlobalFunc.cxx:
```diff
@@ -187,6 +187,7 @@ namespace RooFit {
   RooCmdArg NumCPU(Int_t nCPU, Int_t interleave)   { return RooCmdArg("NumCPU",nCPU,interleave,0,0,0,0,0,0) ; }
   RooCmdArg NewStyle(bool flag) { return RooCmdArg("NewStyle",flag,0,0,0,0,0,0) ; }
   RooCmdArg Parallelize(Int_t nWorkers, bool parallelGradient, bool parallelLikelihood) { return RooCmdArg("Parallelize",nWorkers,parallelGradient,parallelLikelihood,0,0,0,0,0) ; }
+  RooCmdArg LogTimings(bool flag) { return RooCmdArg("LogTimings",flag,0,0,0,0,0,0) ; }
```
You are missing the corresponding declaration in `RooGlobalFunc.h`

guitargeek commented on roofit/multiprocess/res/RooFit/MultiProcess/ProcessTimer.h:
```diff
@@ -0,0 +1,74 @@
+#include <nlohmann/json.hpp>
+using json = nlohmann::json;
+
+using namespace std;
```
```suggestion
```
Remember the C++ rule to never do `using namespace` in header files. Everybody who includes the header will get these using statements too, and things could become ambiguous.

guitargeek commented on roofit/multiprocess/inc/RooFit/MultiProcess/HeatmapAnalyzer.h:
```diff
@@ -0,0 +1,57 @@
+
+class HeatmapAnalyzer {
+public:
+   HeatmapAnalyzer(TString logs_dir);
```
```suggestion
   HeatmapAnalyzer(std::string const& logs_dir);
```
Try to avoid using the old `TString` please

guitargeek commented on roofit/roofitcore/src/RooAbsPdf.cxx:
```diff
@@ -1738,6 +1743,7 @@ RooFitResult* RooAbsPdf::fitTo(RooAbsData& data, const RooLinkedList& cmdList)
   cfg.nWorkers = pc.getInt("nWorkers");
   cfg.parallelGradient = pc.getInt("parallelGradient");
   cfg.parallelLikelihood = pc.getInt("parallelLikelihood");
+  cfg.logTimings = pc.getInt("logTimings");
```
As we discussed on Mattermost, your new feature only works for a `RooSimultaneous`.

It would be better to make that more explicit, sooner than later, so it doesn't cause more confusion.

What about checking if `this` is a RooSimultaneous, and if not you are ignoring `logTimings` and print a warning with `coutW(Minimization)`, telling that LogTimings is not supported for other PDFs? That would be more graceful than writing out empty JSON files/histograms.

guitargeek commented on roofit/multiprocess/src/HeatmapAnalyzer.cxx:
```diff
@@ -0,0 +1,212 @@
+ * and allows for these timings to be written out in json format, one for each process.
+ * This class, the HeatmapAnalyzer, can read these json files and produce a heatmap from
+ * them with partial derivatives on the y-axis, likelihood evaluations on the x-axis, and
+ * time expenditures on the z-axis. This class also contains some convenience functions
```
What's the unit of the time expenditure on the z-axis? Can you write this here pls?

Maybe you can also add a Z-axis label to the histogram to make that clear also to people that are too lazy to read the docs.

Zeff020 replied:
Time expenditure is in the units it was logged in on the ProcessTimer, so I think it does not make sense I think to mention here the units. The HeatmapAnalyzer does not know anything about units but just sees numbers from the log files. I will change the ProcessTimer to make it clear what units timings are logged in, is that alright?

guitargeek replied:
Yes, that's great! Thanks for the clarification

guitargeek commented:
@phsft-bot

guitargeek reviewed (changes requested):
Hi @Zeff020, thanks for the update!

Is the end-to-end test still coming? I guess you can just do a fit with a RooSimultaneous and some Gaussians (even just one). Like what I tried on my computer:
```c++
   // Declare variables x,mean,sigma with associated name, title, initial value and allowed range
   RooRealVar x("x", "x", -10, 10);
   RooRealVar mean("mean", "mean of gaussian", 1, -10, 10);
   RooRealVar sigma("sigma", "width of gaussian", 1, 0.1, 10);

   // Build gaussian pdf in terms of x,mean and sigma
   RooGaussian gauss("gauss", "gaussian PDF", x, mean, sigma);

   sigma.setVal(3);

   RooCategory sample("sample", "sample");
   sample.defineType("cat0");

   RooSimultaneous simPdf{"simPdf", "simPdf", {{"cat0", &gauss}}, sample};
```

Zeff020 commented:
Hi @guitargeek, I've addressed your changes and rebased to master. I've also changed `LogTimings` to `TimingAnalysis` to it is clear that it is different to the existing profiling and allows for a bit more of an in-depth analysis.

guitargeek commented:
@phsft-bot build
